### PR TITLE
Enable MPS support for LitGPT

### PR DESCRIPTION
**Pull request by rasbt**
Re-enables MPS support for LitGPT with an alternative `index_copy_` implementation.

### CODE REVIEW

File: litgpt/lora.py [hunk 1]
```diff
@@ -338,8 +340,17 @@ def zero_pad(self, x: torch.Tensor) -> torch.Tensor:
...
         # embeddings_size is 384 (self.linear.out_features), so that means that we need to pad from 256 to 384 with zeros, but
         # only for key updates (this is where self.lora_ind comes in handy)
-        result = x.new_zeros(*x.shape[:-1], self.linear.out_features)  # (64, 64, 384)
-        return result.index_copy_(dim=-1, index=self.lora_ind, source=x)  # (64, 64, 384)
+
+        if not self.mps_compatibility_mode:
+            result = x.new_zeros(*x.shape[:-1], self.linear.out_features)  # (64, 64, 384)
+            return result.index_copy_(dim=-1, index=self.lora_ind, source=x)  # (64, 64, 384)
+        else:
+            if all(self.enable_lora):
```
Comment (Andrei-Aksionov) [inline litgpt/lora.py:348]: That shouldn't be in the compatibility mode.
In fact, only index copy should be altered with slicing in the compatibility mode.

**Comment by t-vi** [comment]
To my mind, the implementation should not wire a mps compatibility mode through the model but just check the tensor device in index copy.
This is quite a layering leak and does not look like good UX.
If we're concerned about testing, we should bump the litgpt ci  to apple silicon or have a function for the mps impl that we test separately.

**Comment by Andrei-Aksionov** [comment]
> just check the tensor device in index copy.

Agree.

> If we're concerned about testing, we should bump the litgpt ci to apple silicon or have a function for the mps impl that we test separately.

M1 runner for public repos should be free of charge: https://github.com/actions/runner-images/issues/9254

**Comment by t-vi** [comment]
Yeah, we need to switch to M1 anyways because x86 macs don't get PyTorch support anymore (I think the last version was 2.2).

**Comment by rasbt** [comment]
> To my mind, the implementation should not wire a mps compatibility mode through the model but just check the tensor device in index copy.
This is quite a layering leak and does not look like good UX.
If we're concerned about testing, we should bump the litgpt ci to apple silicon or have a function for the mps impl that we test separately.

Yeah, it's a bit tricky. One could do

```python
def batched_index_copy_(t, dim, idx, val):
    """Index copy for batched t, idx, val"""

    if torch.backends.mps.is_available():
        # mps compatible code
    else:
        # orig index_copy_ code
```

but the problem is then what if someone has an MPS machine but deliberately wants to run on CPU?

If you have alternative suggestions for dealing with that, I am happy with other implementation solutions (I also don't like the current one, but it was just something to get fixed before next week before the conference).

**Pull request merged**